tools/internal/parser/text.go

```diff
@@ -148,3 +163,171 @@ func (s Source) forEachRun(classify func(line Source) bool, processBlock func(bl
+//
+// The canonical PSL encoding is a file consisting entirely of valid
+// UTF-8, with no leading BOM or unicode replacement characters. In an
+// effort to report useful errors for common mangling caused by older
+// Windows software, normalizeToUTF8Lines accepts input encoded as
+// UTF-8, UTF-16LE or UTF-16BE, with or without a leading BOM.
+//
+// normalizeToUTF8Lines returns the normalized lines of bs, as well as
+// errors that report deviations from the canonical encoding, if any.
+func normalizeToUTF8Lines(bs []byte) ([]string, []error) {
+	var errs []error
+
+	enc := utf8Transform
+	switch {
+	case bytes.HasPrefix(bs, []byte(bomUTF8)):
+		errs = append(errs, UTF8BOMError{})
+	case bytes.HasPrefix(bs, []byte(bomUTF16BE)):
+		enc = utf16BigEndianTransform
+		errs = append(errs, InvalidEncodingError{"UTF-16BE"})
+	case bytes.HasPrefix(bs, []byte(bomUTF16LE)):
+		enc = utf16LittleEndianTransform
+		errs = append(errs, InvalidEncodingError{"UTF-16LE"})
+	default:
+		enc = guessUTFVariant(bs)
+		switch enc {
+		case utf16BigEndianTransform:
+			errs = append(errs, InvalidEncodingError{"UTF-16BE (guessed)"})
+		case utf16LittleEndianTransform:
+			errs = append(errs, InvalidEncodingError{"UTF-16LE (guessed)"})
+		}
+	}
+
+	bs, err := enc.NewDecoder().Bytes(bs)
+	if err != nil {
+		// The decoder shouldn't error out, if it does we can't really
+		// proceed, just return the errors we've found so far.
+		errs = append(errs, err)
+		return []string{}, errs
+	}
+
+	if len(bs) == 0 {
+		return []string{}, errs
+	}
+
+	ret := strings.Split(string(bs), "\n")
+	for i, line := range ret {
+		// capture source info before we tidy up the line starts/ends,
+		// so that input normalization errors show the problem being
+		// described.
+		//
+		// However, we still provide post-sanitization UTF-8 bytes,
+		// not the raw input. The raw input is unlikely to display
+		// correctly in terminals and logs, and because the unicode
+		// replacement character is a distinctive shape that stands
+		// out, it should provide enough hints as to where any invalid
+		// byte sequences are.
+		src := Source{
+			lineOffset: i,
+			lines:      []string{line},
+		}
+		if strings.ContainsRune(line, utf8.RuneError) {
+			errs = append(errs, InvalidUTF8Error{src})
+		}
+		line, ok := strings.CutSuffix(line, "\r")
+		if ok {
+			ret[i] = line
+			errs = append(errs, DOSNewlineError{src})
+		}
+		if ln := strings.TrimRightFunc(line, unicode.IsSpace); ln != line {
+			line = ln
+			ret[i] = line
+			errs = append(errs, TrailingWhitespaceError{src})
+		}
+		if ln := strings.TrimLeftFunc(line, unicode.IsSpace); ln != line {
+			line = ln
+			ret[i] = line
+			errs = append(errs, LeadingWhitespaceError{src})
+		}
+	}
+
+	return ret, errs
+}
+
+// guessUTFVariant guesses the encoding of bs.
+//
+// Returns the transformer to use on bs, one of utf8Transform,
+// utf16LittleEndianTransform or utf16BigEndianTransform.
+func guessUTFVariant(bs []byte) encoding.Encoding {
```

Comment:
Can you elaborate a bit on the scenarios you're adding this for? Is this just for the odd case where somebody uploads UTF-16 instead?